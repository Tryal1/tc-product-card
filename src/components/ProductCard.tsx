import React, { createContext, CSSProperties } from "react";
import { useProduct } from "../hooks/useProducts";
import {
  Product,
  ProductContextProps,
  onChangeArgs,
  InitialValue,
  ProductCardHandlers,
} from "../interfaces/interfaces";
import styles from "../styles/styles.module.css";

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children?: (arg: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValue?: InitialValue;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValue,
}: Props) => {
  const { counter, increasBy, maxCount, isMaxReached, reset } = useProduct({
    onChange,
    product,
    value,
    initialValue,
  });

  return (
    <Provider
      value={{
        counter,
        increasBy,
        product,
        maxCount,
      }}
    >
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children &&
          children({
            count: counter,
            isMaxReached,
            maxCount: initialValue?.maxCount,
            product,
            increasBy,
            reset,
          })}
      </div>
    </Provider>
  );
};
