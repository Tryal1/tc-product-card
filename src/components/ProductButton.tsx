import React, { CSSProperties, useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";

export interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductButton = ({ className, style }: Props) => {
  const { increasBy, counter, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(
    () => !!maxCount && counter === maxCount,
    [counter, maxCount]
  );

  return (
    <div style={style} className={`${styles.buttonsContainer} ${className}`}>
      <button onClick={() => increasBy(-1)} className={styles.buttonMinus}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        onClick={() => increasBy(1)}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disa}`}
      >
        +
      </button>
    </div>
  );
};
