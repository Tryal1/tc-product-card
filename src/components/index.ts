
import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductCardHOCProps } from '../interfaces/interfaces';

import { ProductButton } from './ProductButton';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';

export { ProductButton } from './ProductButton';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';


export const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButton,
})


export default ProductCard;