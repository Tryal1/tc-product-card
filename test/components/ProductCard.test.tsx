import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product1 } from '../data/producto';

const { act } = renderer;

describe('ProductImage', () => {
  test('Debe de mostrar el componente correctamente con la Imagen personalizado', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => <h1>Product Card</h1>}
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Debe Incrementar el contador', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {({ count, increasBy }) => (
          <>
            <h1>product card</h1>
            <span>{count}</span>
            <button onClick={() => increasBy(1)}></button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();

    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
