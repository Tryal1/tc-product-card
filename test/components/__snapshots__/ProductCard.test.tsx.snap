// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe Incrementar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    product card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`ProductImage Debe de mostrar el componente correctamente con la Imagen personalizado 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
</div>
`;
